Support Gitlab instances for source repositories

cmelone opened the pull request:
This PR enables syncing commits (and pipeline statuses) between repositories on two Gitlab instances.

Modifications were made to several parts of the codebase:
- global config options are more verbose to accommodate Gitlab and Github both being options for the source
- there are now two Gitlab client classes (for dest and src). the logic is different and consolidating them would have a negative effect on readability, imo (open to disagreement though)
- there are also two Gitlab handler classes for similar reasons
- many new route handlers in `src/hubcast/web/gitlab/routes.py`

Feedback would be helpful on software arch decisions and potential consolidation of functionality.

closes #115

outstanding issues (see comments below):

- [ ] can't post pipeline status to forked repos
- [ ] bot user for source gitlab instance?
- [ ] naming of handler and client classes for src/dest

TODOs after this PR:

- [ ] bot/commenting functionality
- [ ] draft PR sync conditional

cmelone committed:
Support Gitlab instances for source repositories

cmelone commented:
re: bot setup if using gitlab as the source repo

Gitlab does not provide a similar mechanism to Github apps, where a collection of permissions/webhooks can be "installed" to a repo.

From what I've seen in other examples ([codecov](https://docs.codecov.com/docs/team-bot), [jenkins](https://plugins.jenkins.io/gitlab-plugin/)), the workaround is:

Instruct Hubcast users to create a "bot" account on the source Github instance. The account should have access at the `Developer` role.
   - An access token with `api` and `read_repository` access would be provided as an env variable to hubcast along with the bot username.
   - The bot user could be given access to many Gitlab groups or individual repositories by the administrators of those entities.
   - The bot would set the status of pipelines and comment on PRs.

Downsides:
- each repository would need to configure a webhook to create events upon pushes, comments, and merge request changes
   - similar to the way we handle this for the destination, we could handle the creation of this webhook upon setup/register of the repository somehow

cmelone committed:
small ordering change

cmelone commented:
the handlers and clients have names like `GitLabSrcClientFactory` and `GitLabSrcHandler`. I haven't changed the destination equivalents yet, or GitHub classes if there's a more attractive option available

alecbcs committed:
Update README.md for GitLab -> GitLab

alecbcs commented on src/hubcast/account_map/abc.py:
```diff
@@ -8,9 +8,8 @@ class AccountMap(ABC):
         """
-        Return the coorisponding gitlab_user for a given github_user if
-        one exists.
+        Return the corresponding gitlab_user for a given src_user if one exists.
```
```suggestion
        Return the corresponding dest_forge_user for a given src_forge_user if one exists.
```

cmelone commented on src/hubcast/account_map/abc.py:
```diff
@@ -8,9 +8,8 @@ class AccountMap(ABC):
 
     @abstractmethod
-    def __call__(self, github_user: str) -> Union[str, None]:
+    def __call__(self, src_user: str) -> Union[str, None]:
```
yes, that sounds good to me! I'll look for other references that can be generalized as well

cmelone committed:
fix syncing typo

cmelone committed:
Standardize on 'forge' language

Co-authored-by: Caetano Melone <melone1@llnl.gov>
Co-authored-by: Alec Scott <scott112@llnl.gov>

cmelone committed:
define repo config filepath based on source forge

1

cmelone committed:
fix stray traceback

cmelone committed:
add authentication to destination repo ls_remote to account for potential privateness

cmelone committed:
add support for private source repositories

- checks if a source repository is private and authenticates if
  necessary
- private forks are skipped as we cannot access them
- see https://github.com/LLNL/hubcast/pull/189 for the github equivalent
  implementation

cmelone committed:
enforce unique have_shas during sync (see #199)

cmelone committed:
update gitlab->gitlab sync for compatibility with gl dest impersonation tokens (#188)

cmelone committed:
update authed repo actions for latest #189 changes

cmelone commented:
this PR has veered far off the mainline. closing this in favor of a newer that I will submit soon
